src/main/java/thePackmaster/powers/pixiepack/UpgradedFastHandsPower.java

```diff
@@ -0,0 +1,51 @@
+package thePackmaster.powers.pixiepack;
+
+import com.evacipated.cardcrawl.mod.stslib.powers.interfaces.NonStackablePower;
+import com.megacrit.cardcrawl.actions.common.DrawCardAction;
+import com.megacrit.cardcrawl.cards.AbstractCard;
+import com.megacrit.cardcrawl.core.AbstractCreature;
+import com.megacrit.cardcrawl.core.CardCrawlGame;
+import com.megacrit.cardcrawl.dungeons.AbstractDungeon;
+import com.megacrit.cardcrawl.localization.PowerStrings;
+import com.megacrit.cardcrawl.monsters.AbstractMonster;
+import thePackmaster.SpireAnniversary5Mod;
+import thePackmaster.packs.PixiePack;
+import thePackmaster.powers.AbstractPackmasterPower;
+
+public class UpgradedFastHandsPower extends AbstractPackmasterPower {
+    public static final String POWER_ID = SpireAnniversary5Mod.makeID("UpgradedFastHandsPower");
+    private static final PowerStrings powerStrings = CardCrawlGame.languagePack.getPowerStrings(POWER_ID);
+    public static final String NAME = powerStrings.NAME;
+    public static final String[] DESCRIPTIONS = powerStrings.DESCRIPTIONS;
+
+    int costReductionAmount;
+
+    public UpgradedFastHandsPower(AbstractCreature owner, int amount) {
+        super(POWER_ID, NAME, PowerType.BUFF,false, owner, amount);
+    }
+
+    @Override
+    public void onCardDraw(AbstractCard card) {
+        super.onCardDraw(card);
+        if (costReductionAmount > 0)
+        {
+            costReductionAmount--;
+            if (card.costForTurn > 0) card.setCostForTurn(card.costForTurn - 1);
+        }
+    }
+
+    @Override
+    public void onPlayCard(AbstractCard card, AbstractMonster m) {
+        super.onPlayCard(card, m);
+        if (PixiePack.isForeign(card))
+        {
+            AbstractDungeon.actionManager.addToTop(new DrawCardAction(amount));
```

Comment:
There's a different way you can do this that avoids the need for what you're doing with `costReductionAmount` and `onCardDraw`, by using the built-in follow up action for `DrawCardAction`.

Try something like this out (haven't tried to compile this but should be roughly correct):
```
        AbstractDungeon.actionManager.addToTop(new DrawCardAction(amount, new AbstractGameAction() {
            @Override
            public void update() {
                for (AbstractCard card : DrawCardAction.drawnCards) {
                    if (card.costForTurn > 0) card.setCostForTurn(card.costForTurn - 1);
                }
                this.isDone = true;
            }
        }));
```

Then you can entirely delete `onCardDraw` and `costReductionAmount`.

Reply:
Fixed, code did indeed just work!